### PR TITLE
security: add certificate rotation timestamp and expiry days metrics

### LINKED ISSUE #167380: security: add certificate rotation timestamp and expiry metrics
**Is your feature request related to a problem? Please describe.**

Operators currently lack visibility into certificate lifecycle events. There is no easy way to monitor when certificates were last rotated or how close they are to expiring, making it difficult to set up proactive alerts.

**Describe the solution you'd like**

Add the following certificate lifecycle metrics:

1. **Rotation Timestamps**
   `security.certificate.last_rotation.{ca, node, ui, ...}` — Unix timestamp of the last successful cert reload where the certificate actually changed.

2. **Days Until Expiry**
   `security.certificate.expiry_days.{ca, node, ...}` — days until certificate expiration.

**Describe alternatives you've considered**

Operators can manually inspect certificate files or parse logs, but this is error-prone and doesn't integrate with standard monitoring/alerting pipelines.

**Additional context**

These metrics will help operators monitor cert health, set up alerts before expiry, and verify that rotations are happening as expected.

Epic CC-34831
Epic CRDB-59291

Jira issue: CRDB-62423

### PR DESCRIPTION
**Pull request by souravcrl**
Operators lacked visibility into certificate lifecycle events — there was no easy way to monitor when certificates were last rotated or how many days remain until they expire. This made it difficult to set up proactive alerting.

Add two new families of certificate metrics:

1. Rotation timestamps (`security.certificate.last_rotation.*`): Gauge recording the Unix timestamp of the last cert reload where the certificate content actually changed. Remains 0 until the first rotation after process start. Covers ca, client-ca, ui-ca, ca-client-tenant, node, node-client, ui, and client-tenant.

2. Days until expiry (`security.certificate.expiry_days.*`): FunctionalGauge reporting ceil(ttl_seconds / 86400). Returns 0 when the certificate is expired, missing, or has an error. Covers the same set of certificate types.

Rotation detection works by comparing old vs new FileContents in LoadCertificates before the swap, so a no-op reload (same cert on disk) does not update the timestamp.

Resolves: #167380
Epic: CRDB-59291

Release note (ops change): Added new certificate lifecycle metrics: `security.certificate.last_rotation.*` reports the Unix timestamp of the most recent certificate rotation, and
`security.certificate.expiry_days.*` reports the number of days remaining until each certificate expires. These enable operators to set up monitoring alerts for certificate health.

### CODE REVIEW

File: pkg/security/certificate_metrics.go [hunk 1]
```diff
@@ -217,6 +242,138 @@ var (
 		LabeledName:  "security.certificate.ttl",
 		StaticLabels: metric.MakeLabelPairs(metric.LabelCertificateType, "client-tenant"),
 	}
+
+	// Rotation timestamp metadata.
+	metaCALastRotation = metric.Metadata{
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics.go:247]: Missing Visibility and Category on all 16 new metric metadata definitions. The existing expiration metrics in the same file set Visibility: metric.Metadata_ESSENTIAL and Category: metric.Metadata_EXPIRATIONS.

File: pkg/security/certificate_metrics.go [hunk 2]
```diff
@@ -217,6 +242,138 @@ var (
...
+		Name:         "security.certificate.last_rotation.client-tenant",
+		Help:         "Unix timestamp of the last tenant client certificate rotation. 0 means no rotation since process start.",
+		Measurement:  "Certificate Last Rotation",
+		Unit:         metric.Unit_TIMESTAMP_SEC,
+		LabeledName:  "security.certificate.last_rotation",
+		StaticLabels: metric.MakeLabelPairs(metric.LabelCertificateType, "client-tenant"),
+	}
+
+	// Days-until-expiry metadata.
+	metaCAExpiryDays = metric.Metadata{
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics.go:337]: security.certificate.expiry_days.* duplicates the existing security.certificate.ttl.* metrics. Both measure remaining certificate lifetime TTL in seconds, expiry-days as ceil(seconds/86400). PromQL can trivially derive days from seconds.

File: pkg/security/certificate_metrics.go [hunk 3]
```diff
@@ -254,6 +411,24 @@ func ttlGauge(
 	})
 }
 
+const secondsPerDay = 86400
+
+func expiryDaysGauge(
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics.go:468]: Add a function comment to the function, especially given the non-obvious math.Ceil rounding.

File: pkg/security/certificate_metrics.go [hunk 4]
```diff
@@ -217,6 +242,138 @@ var (
...
+		Help:         "Unix timestamp of the last tenant client certificate rotation. 0 means no rotation since process start.",
+		Measurement:  "Certificate Last Rotation",
+		Unit:         metric.Unit_TIMESTAMP_SEC,
+		LabeledName:  "security.certificate.last_rotation",
+		StaticLabels: metric.MakeLabelPairs(metric.LabelCertificateType, "client-tenant"),
+	}
+
+	// Days-until-expiry metadata.
+	metaCAExpiryDays = metric.Metadata{
+		Name:         "security.certificate.expiry_days.ca",
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics.go:338]: expiry_days uses an underscore where the rest of the metric namespace uses dots (security.certificate.expiration.ca, security.certificate.ttl.ca). Consider security.certificate.expiry.days.* for consistency.

File: pkg/security/certificate_metrics_test.go [hunk 5]
```diff
@@ -106,6 +107,199 @@ func TestMetricsValues(t *testing.T) {
 
 }
 
+// TestExpiryDaysValues verifies that the days-until-expiry metrics report
+// the correct number of days for each certificate type.
+func TestExpiryDaysValues(t *testing.T) {
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics_test.go:113]: All 8 cert types produce an expected value of 1 day:                         
{"CA Expiry Days", 1, metrics.CAExpiryDays.Value()},                                                  
{"Client CA Expiry Days", 1, metrics.ClientCAExpiryDays.Value()},                                     
// ... all 1

If the code accidentally wired CAExpiryDays to read the node cert (or any other cert), the test still passes because they all return 1.                                               
Fix: Use expirations that produce distinct day counts per cert type. For example, set now to 0 and give CA an expiration of 1 day, ClientCA 2 days, Node 5 days, etc. Then each assertion has a unique expected value, and a wiring mixup would fail.

Same for TestRotationTimestampsAllCertTypes

**Review by sanchit-CRL — COMMENTED** [review]
Test coverage gaps, missing edge cases for expiry-days and rotation metrics                          
A few scenarios documented in the metric Help text ("0 means expired, no certificate or error") aren't covered by the current tests:                                                                        

1. Expired cert returns 0, no test exercises the sec < 0 branch in expiryDaysGauge. Set now past the cert expiration and assert 0.
2. Errored cert returns 0 — write garbage to a cert file so CertInfo.Error is set, assert the expiry-days gauge reads 0.
3. Missing cert returns 0 — create a CertificateManager with only the required certs (CA + node) and verify optional cert metrics (UI CA, tenant CA, etc.) return 0.

File: pkg/security/certificate_metrics.go (hunk 1 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics.go:247]: done

File: pkg/security/certificate_metrics.go (hunk 2 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics.go:337]: I had discussed this earlier on internal channels but it was done so we don't have to additionally write a query to do the transformation and since these are aggregate, they won't be accurate. I think data for days till expiry makes more sense here since operators need that to monitor jobs or trigger certificate rotation.

File: pkg/security/certificate_metrics.go (hunk 4 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics.go:338]: For the advanced metrics https://www.cockroachlabs.com/docs/stable/essential-metrics-advanced#row-level-ttl I think we are following `.` for a separate entity like `row_level_ttl` is a separate entity from `currently_paused` but both the word `row_level_ttl` has significance in metrics context hence it is a single entity. I maybe wrong on this but this is what I understand from the examples.

File: pkg/security/certificate_metrics_test.go (hunk 5 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics_test.go:113]: done

**Comment by sanchit-CRL** [comment]
> Test coverage gaps, missing edge cases for expiry-days and rotation metrics A few scenarios documented in the metric Help text ("0 means expired, no certificate or error") aren't covered by the current tests:
> 
> 1. Expired cert returns 0, no test exercises the sec < 0 branch in expiryDaysGauge. Set now past the cert expiration and assert 0.
> 2. Errored cert returns 0 — write garbage to a cert file so CertInfo.Error is set, assert the expiry-days gauge reads 0.
> 3. Missing cert returns 0 — create a CertificateManager with only the required certs (CA + node) and verify optional cert metrics (UI CA, tenant CA, etc.) return 0.

Unaddressed comment

**Comment by souravcrl** [comment]
> > Test coverage gaps, missing edge cases for expiry-days and rotation metrics A few scenarios documented in the metric Help text ("0 means expired, no certificate or error") aren't covered by the current tests:
> > 
> > 1. Expired cert returns 0, no test exercises the sec < 0 branch in expiryDaysGauge. Set now past the cert expiration and assert 0.
> > 2. Errored cert returns 0 — write garbage to a cert file so CertInfo.Error is set, assert the expiry-days gauge reads 0.
> > 3. Missing cert returns 0 — create a CertificateManager with only the required certs (CA + node) and verify optional cert metrics (UI CA, tenant CA, etc.) return 0.
> 
> Unaddressed comment

I have addressed it now.

File: pkg/internal/metricscan/metric_owners.yaml [hunk 6]
```diff
@@ -890,6 +889,24 @@ owners:
   security_certificate_expiration_node_client: cockroachdb/security-engineering
   security_certificate_expiration_ui: cockroachdb/security-engineering
   security_certificate_expiration_ui_ca: cockroachdb/security-engineering
+  security_certificate_expiry_days: cockroachdb/security-engineering
```
Comment (sanchit-CRL) [inline pkg/internal/metricscan/metric_owners.yaml]: should this be product-security ?

File: pkg/security/certificate_metrics_test.go [hunk 7]
```diff
@@ -106,6 +107,352 @@ func TestMetricsValues(t *testing.T) {
...
+	require.Greater(t, metrics.NodeExpiryDays.Value(), int64(0), "node should report > 0 days")
+	require.Greater(t, metrics.NodeClientExpiryDays.Value(), int64(0), "node-client should report > 0 days")
+	// Missing optional certs should report 0.
+	require.Equal(t, int64(0), metrics.UICAExpiryDays.Value(), "missing UI CA should report 0 days")
+	require.Equal(t, int64(0), metrics.ClientCAExpiryDays.Value(), "missing client CA should report 0 days")
+	require.Equal(t, int64(0), metrics.TenantCAExpiryDays.Value(), "missing tenant CA should report 0 days")
+	require.Equal(t, int64(0), metrics.UIExpiryDays.Value(), "missing UI cert should report 0 days")
+	require.Equal(t, int64(0), metrics.TenantExpiryDays.Value(), "missing tenant cert should report 0 days")
+}
+
```
Comment (sanchit-CRL) [inline pkg/security/certificate_metrics_test.go:455]: TestCertificateMetricsReloadRace reads existing metrics concurrently with LoadCertificates() to catch data races. The new ExpiryDays and LastRotation metrics are not included in this race test. This is the exact class of bug that test was designed to catch.

File: pkg/security/certificate_metrics.go (hunk 2 above)
Comment (sanchit-CRL) [reply pkg/security/certificate_metrics.go:337]: Can you mention the internal thread so that I gain more context ?

File: pkg/internal/metricscan/metric_owners.yaml (hunk 6 above)
Comment (souravcrl) [reply pkg/internal/metricscan/metric_owners.yaml]: I am not sure, I have changed it to product-security.

File: pkg/security/certificate_metrics_test.go (hunk 7 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics_test.go:455]: updated.

**Comment by souravcrl** [comment]
> Can you mention the internal thread so that I gain more context ?

https://cockroachlabs.slack.com/archives/C06G0FR568N/p1774867747509739?thread_ts=1774665064.932229&cid=C06G0FR568N

File: pkg/security/certificate_metrics.go (hunk 2 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics.go:337]: https://cockroachlabs.slack.com/archives/C06G0FR568N/p1774867747509739?thread_ts=1774665064.932229&cid=C06G0FR568N

File: pkg/internal/metricscan/metric_owners.yaml (hunk 6 above)
Comment (souravcrl) [reply pkg/internal/metricscan/metric_owners.yaml]: Ok had to revert it. 
So the team is determined by CODEOWNERS file. The code generator overwrites product-security back to security-engineering because pkg/security/ is owned by cockroachdb/security-engineering in CODEOWNERS. Currently the `pkg/security/` is owned cockroachdb/security-engineering hence this mapping is currently not possible.

File: pkg/security/certificate_metrics_test.go (hunk 7 above)
Comment (souravcrl) [reply pkg/security/certificate_metrics_test.go:455]: I hadn't touched this test because I wasn't aware of it. This test is maintained by obs I think.

**Review by sanchit-CRL — APPROVED** [review]
LGTM

**Comment by souravcrl** [comment]
TFTR!

/trunk merge
